### DISCUSSION TITLE
Error message generated when activating NFS other server Alternative 2

### DISCUSSION THREAD

**Discussion by Sagnasty** [rern/rAudio#1308] [Fixed Issues]
Running rAudio 20230522

In an attempt to activate NFS other server Alternative 2, pressing the System / Shared data client button generates the following error:

![image](https://github.com/rern/rAudio/assets/76575081/45b32eed-17a4-47ca-b65c-c713547567a1)

Note: that I have not activated the rAudioServer as viewed from the features screen:

![image](https://github.com/rern/rAudio/assets/76575081/d7e21241-63b2-4af9-92bb-d739187c95bc)

FEATURES-DATA

{
	"autoplay": false,
	"autoplayconf": false,
	"brightness": false,
	"camilladsp": false,
	"camilladspconf": {
		"REFRESH": 100
	},
	"equalizer": false,
	"hostapd": false,
	"hostapdconf": {
		"IP": "192.168.5.1",
		"PASSPHRASE": "raudioap"
	},
	"hostip": "192.168.1.10",
	"hostname": "rAudioHome",
	"httpd": false,
	"ipsub": "192.168.1.",
	"latest": false,
	"lcd": false,
	"localbrowser": true,
	"localbrowserconf": false,
	"login": false,
	"lyrics": true,
	"lyricsconf": {
		"URL": "https:/",
		"START": "<",
		"END": "</div>",
		"EMBEDDED": true
	},
	"multiraudio": false,
	"multiraudioconf": false,
	"nfsconnected": false,
	"nfsserver": false,
	"nosound": false,
	"page": "features",
	"scrobble": false,
	"scrobbleconf": false,
	"scrobblekey": false,
	"shairport-sync": false,
	"shareddata": false,
	"smb": false,
	"smbconf": {
		"SD": true,
		"USB": true
	},
	"snapclient": false,
	"snapclientactive": false,
	"snapclientconf": {
		"LATENCY": false
	},
	"snapserver": false,
	"snapserveractive": false,
	"spotifyd": false,
	"spotifyredirect": "https://rern.github.io/raudio/spotify",
	"spotifytoken": false,
	"state": "stop",
	"stoptimer": false,
	"stoptimerconf": false,
	"upmpdcli": false,
	"upmpdcliconf": {
		"OWNQUEUE": false
	},
	"wlan": false,
	"wlanconnected": false
}

3 x Raspberry Pi 3B+ Rev 1.3 > HiFiBerry DAC+ Pro
1 x Raspberry Pi 3B Rev 1.2 > HiFiBerry DAC+ Pro
Synology DS718+

Jim

**Comment by ceejayemm** [comment]
I get this too with 20230522 on RPi£A/3B+ but I DO (or did, prior to this upgrade) have Shared Data activated on a separate CIFS server (not via rAudio Server).

df -h seems to show CIFS server share is mounted (see last line):

rAudio-1:~  #  df -h
Filesystem                                 Size  Used Avail Use% Mounted on
/dev/root                                   15G  2.9G   12G  21% /
devtmpfs                                   191M     0  191M   0% /dev
tmpfs                                      225M   28K  225M   1% /dev/shm
tmpfs                                       90M  656K   90M   1% /run
tmpfs                                      225M  4.0K  225M   1% /tmp
/dev/mmcblk0p1                              99M   46M   53M  47% /boot
192.168.86.250:/export/music/iTunes/Music  3.6T  995G  2.7T  28% /mnt/MPD/NAS/NAS_Music
//192.168.86.250/raudio                    3.6T  995G  2.7T  28% /mnt/MPD/NAS/data

and a directory listing also seems to show the symlinks still in place:

rAudio-1:~  #  ls -lrt /srv/http/data
total 20
drwxr-xr-x 4 http http 4096 Jan 22 13:40 camilladsp
lrwxrwxrwx 1 http http    8 Jan 22 13:41 shm -> /dev/shm
lrwxrwxrwx 1 http http   25 Mar  8 12:15 audiocd -> /mnt/MPD/NAS/data/audiocd
lrwxrwxrwx 1 http http   27 Mar  8 12:15 bookmarks -> /mnt/MPD/NAS/data/bookmarks
lrwxrwxrwx 1 http http   24 Mar  8 12:15 lyrics -> /mnt/MPD/NAS/data/lyrics
lrwxrwxrwx 1 http http   21 Mar  8 12:15 mpd -> /mnt/MPD/NAS/data/mpd
lrwxrwxrwx 1 http http   27 Mar  8 12:15 playlists -> /mnt/MPD/NAS/data/playlists
lrwxrwxrwx 1 http http   26 Mar  8 12:15 webradio -> /mnt/MPD/NAS/data/webradio
drwxr-xr-x 7 http http 4096 May  4 09:27 backup
drwxr-xr-x 3 http http 4096 May 22 12:19 mpdconf
drwxr-xr-x 2 http http 4096 May 22 17:30 system
drwxr-xr-x 2 http http 4096 May 22 17:31 addons

Chris

**Comment by ceejayemm** [comment]
Still an issue with 20230523

Chris

**Reply by Sagnasty** [reply]
Same here.

Based on the fixed list from the releases page, I assumed that rern hasn't gotten to this issue yet.

![image](https://github.com/rern/rAudio/assets/76575081/edc51df6-4407-4c91-af36-ef778f703008)

**Comment by ceejayemm** [comment]
There still doesn't seem to be a fix for this in 20230525 but I seem to have found a workaround (with another possible separate issue but more about that below):

- go to Features / Others and ENABLE 'Server rAudio'
- let it do its thing 
- disable this setting again

You can then go to System / Data and Settings and enable and configure Shared Data (CIFS or NFS) and point it to your alternative Shared Data source.

HOWEVER

on completion of the above I then get a new error:

```
Errors: Unexpected non-whitespace character after JSON at position 1073 
{ 
  "player"       : "mpd"
, "btreceiver"   : false
, "card"         : 0
, "consume"      : false
, "control"      : "Digital"
, "counts"       : {
  "album": 336,
  "albumartist": 131,
  "artist": 638,
  "composer": 1052,
  "conductor": 2,
  "dabradio": 0,
  "date": 162,
  "genre": 38,
  "latest": 0,
  "nas": 1,
  "playlists": 2,
  "sd": 0,
  "song": 5089,
  "usb": 0,
  "webradio": 24
}
, "icon"         : ""
, "librandom"    : false
, "lyrics"       : false
, "relays"       : false
, "relayson"     : false
, "scrobble"     : false
, "shareddata"   : true
, "snapclient"   : false
, "stoptimer"    : false
, "updateaddons" : false
, "updating_db"  : false
, "updatingdab"  : false
, "volume"       : 64
, "volumemute"   : 0
, "webradio"     : false
, "display"          : false
, "audiocd"          : false
, "camilladsp"       : false
, "color"            : ""
, "dabradio"         : false
, "equalizer"        : false
, "multiraudio"      : true
, "relays"           : false
, "screenoff"        : false
, "snapclient"       : false
, "volumenone"       : false
}
X, "repeat" : false               <-- ERROR HERE
, "random" : false
, "single" : false
, "file"      : ""
, "pllength"  : 0
, "song"      : 0
, "state"     : "stop"
, "timestamp" : 1685177496398
, "coverart" : ""
, "hostname" : "raoffice.cmhomenet.lan"
, "ip"       : "192.168.86.31"
, "login" : false }```

which only seems appears on a reboot or restart of the system.  This can also be (temporarily) worked around by changing the url to:

http://<system ip number>/settings.php?p=system

and you now have a fully (?) working system until the next restart. EDIT Spoke too soon.  This 'workaround' only works whie ever you stay in the Settings area.  When you try to exit Settings to return to the main URL the error above reappears. So may be NOT a fix after all

This seems to work, as far as it goes, with i20230522 and updates 20230523 and 20230525, at least for me.

Chris

**Reply by Sagnasty** [reply]
Seems that the workaround is almost as painful as the reported error.

It is good that you managed to get some additional detail on this - hopefully your details will make a permanent solution easier to achieve.

**Reply by rern** [reply]
Try: `> /etc/exports` then setup Shared Data again

**Comment by Sagnasty** [comment]
Works perfectly - tested on Raspberry Pi 3B+ Rev 1.3 and Raspberry Pi 3B Rev 1.2

**Comment by ceejayemm** [comment]
@rern 

Your command seems to have created an empty file /etc/exports - is that what you expected ?  Unfortunately fro me this HASN'T fixed this problem, giving the same error as noted this morning, even following  reboot:

Errors: Unexpected non-whitespace character after JSON at position 1073 
{ 
  "player"       : "mpd"
, "btreceiver"   : false
, "card"         : 0
, "consume"      : false
, "control"      : "Digital"
, "counts"       : {
  "album": 336,
  "albumartist": 131,
  "artist": 638,
  "composer": 1052,
  "conductor": 2,
  "dabradio": 0,
  "date": 162,
  "genre": 38,
  "latest": 0,
  "nas": 1,
  "playlists": 2,
  "sd": 0,
  "song": 5089,
  "usb": 0,
  "webradio": 24
}
, "icon"         : ""
, "librandom"    : false
, "lyrics"       : false
, "relays"       : false
, "relayson"     : false
, "scrobble"     : false
, "shareddata"   : true
, "snapclient"   : false
, "stoptimer"    : false
, "updateaddons" : false
, "updating_db"  : false
, "updatingdab"  : false
, "volume"       : 64
, "volumemute"   : 0
, "webradio"     : false
, "display"          : false
, "audiocd"          : false
, "camilladsp"       : false
, "color"            : ""
, "dabradio"         : false
, "equalizer"        : false
, "multiraudio"      : true
, "relays"           : false
, "screenoff"        : false
, "snapclient"       : false
, "volumenone"       : false
}
X, "repeat" : false
, "random" : false
, "single" : false
, "file"      : ""
, "pllength"  : 0
, "song"      : 0
, "state"     : "stop"
, "timestamp" : 1685205409602
, "coverart" : ""
, "hostname" : "raoffice.cmhomenet.lan"
, "ip"       : "192.168.86.31"
, "login" : false }```

If it makes any difference I an trying to use a CIFS share to an external OMV server which has been used previously without any issue.

Chris

**Comment by Sagnasty** [comment]
Testing this again after applying rAudio 20230528 update.

After successfully creating/mounting  the NFS other server Alternative 2, the slider bar automatically returns back to the OFF position.

So there is no way to turn off the shared data and restore the local data using the shared data slider bar with the rAudio 20230528 update.

**Reply by rern** [reply]
Shared Data status - enabled:
- `/srv/http/data/mpd` is symlink
- `nfs-server` is inactive.

Get both:
```sh
ls -l /srv/http/data
systemctl status nfs-server
```

**Reply by Sagnasty** [reply]
@rern here are the results:
[Shared data disabled.txt](https://github.com/rern/rAudio/files/11599742/Shared.data.disabled.txt)
[Shared data enabled.txt](https://github.com/rern/rAudio/files/11599743/Shared.data.enabled.txt)

**Reply by Sagnasty** [reply]
Problem still exists with rAudio 20230603

3 x Raspberry Pi 3B+ Rev 1.3 > HiFiBerry DAC+ Pro
1 x Raspberry Pi 3B Rev 1.2 > HiFiBerry DAC+ Pro
Synology DS718+

Jim

**Comment by ceejayemm** [comment]
@rern 

I still cannot get this to work, despite updating to 20230528.  I initially thought it was because I was updating a system based on an earlier image (i20230312 I think) so I flashed a new SD card using the i20230522 image, applied the 20230528 update and manually reconfigured the new installation (so no restored backup from a previous version). I re-enabled the Shared Data setup using a new CIFS share and like @Sagnasty above I can see the share is mounted (but the SD card boot mount details disappear? from the list but it is mounted in fstab)  The slider for Shared Data also returns to the Off position and I get exactly the same JSON as previously (X, "repeat" : false) but the line number is now 1069.

Please let me know if you want me to try anything else.

Chris

**Reply by rern** [reply]
Hi Chris,

Local storages, SD and USB, will be hidden once Shared Data was enabled to exclude them from database.

Please try the latest image with a new share directory.

**Reply by ceejayemm** [reply]
OK, I have flashed a new SD card with i20230521 for my RPi 3B (not Plus version, Wired LAN) and used it to boot the RPi 3B. Once started I manually configured the system: 

- to use my IQaudIO DAC Pro, 
- set the timezone to Europe/London, 
- enabled No Volume, enabled FFmpeg, 
- disabled Browser on RPi (which always seems enabled by default these days). 
- Server rAudio was disabled by default and I left it alone. 

At this point the System page still shows the version as 20230310 (despite using the latest image).  I then 

- applied the latest update (20230528)
- enabled Shared Data using a new CIFS share from my OMV NAS. The symlinks were created as expected and the Paradise Radio radio stations AND station icons were copied to the Shared Data location.
- after the initial setup of the Shared Data area the slider however returned to Disabled (but the Shared Data remained setup and usable).
- mounted my NAS based music library using NFS
- the Music Library was detected and a full database update was done (although I hadn't set anything to do this). Once complete I could play albums from the Library as I would expect.
- I added 2 new radio stations which were updated to the Shared Data area BUT I could NOT add any artwork as Change Cover Art did NOT produce any dialogue box for this to be done - but the stations were usable, just no station icons.
- trying to change Playback screen options was hit and miss, I can turn bars on and off but I can't turn off the Time display for example (or Buttons or Artwork). I haven't as yet tried changing any Library or Playlist page options.
- after a reboot, the system version now showed version to be 20230528

So I seem to have got a bit further with no JSON errors at this point.

I then flashed a new SD card in exactly the same way to use with my RPi 3A (wireless) including a 'wifi' file as I normally doe to enable my wifi setup. I followed the same initial setup as above and only updated to 20230528 after the initial configuration was complete. Again no apparent JSON errors this time. I updated to 20230528, again no problem, and enable Shared Data using the same CIFS shared used above. The shared music library was found and was usable but updated system has all the same issues as noted above.

So, fingers crossed, we might be making progress but there obviously still some things wrong with this setup. It seems that configuring the system BEFORE applying the update in this case seems (perhaps ?) to stop the JSON errors I have had in the past. I have 2 (nearly) usable systems but there are still some things need resolving before I can say they are fully back to normal.

Chris

**Comment by ceejayemm** [comment]
@rern 

Now that we seem to resolved the JSON errors with the Shared Data setup can we get back to how it works ?  In previous versions of rAudio when the Shared Data option was selected:

1. A configuration dialogue box (CIFS / NFS /rAudio) was displayed. 
2. Once the necessary information was entered the relevant share was mounted, 
3. rAudio did its thing to move various folder/files to the Shared Data area
4. The Shared Data slider was shown as Enabled and a Cog icon was displayed to allow the configuration to be modified if necessary
5. The Shared Data setup could be disabled / removed by sliding the slider back to disabled.

In the current version, steps 1 to 3 still seem to happen and Shared Data (mostly) seems to work.  The Shared Data slider however stays as Disabled, no Cog icon appears and the only way I have found to change or remove the Shared Data setup is to 'Forget' the folder mounted by the initial Shared Data setup and start again.

Is this the way its meant to work now ?  Is it possible to get the original processes back again please ?

Thanks

Chris

**Reply by rern** [reply]
Hi Chris,

Try:
`sed -i 's/! $nfsserver/$nfsserver == false/' $dirsettings/system-data.sh`

**Comment by Sagnasty** [comment]
Hello Rern,
Running rAudio 20230603
Tried this command with NFS shared server and it is working fine (steps 1 to 5 as @ceejayemm described above).
Now it works the same as the original process did in earlier versions. 
Jim

**Comment by ceejayemm** [comment]
@rern 
Yes, this seems to have cured the problem outlined above for a CIFS based share too. Unfortunately however somewhere along the way it seems I inadvertently enabled the rAudio Server as well and although I turned it off again before testing this change I now have the original JSON error back again. This occurs when I exit the System (or any other set up screen) to go back to the playback screen.  The error is:

Errors: Unexpected non-whitespace character after JSON at position 867 
{ 
  "player"       : "mpd"
, "btreceiver"   : false
, "card"         : 0
, "consume"      : false
, "control"      : "Digital"
, "counts"       : {
  "playlists" : 0
, "webradio"  : 4
}
, "icon"         : ""
, "librandom"    : false
, "lyrics"       : false
, "relays"       : false
, "relayson"     : false
, "scrobble"     : false
, "shareddata"   : true
, "snapclient"   : false
, "stoptimer"    : false
, "updateaddons" : true
, "updating_db"  : false
, "updatingdab"  : false
, "volume"       : 100
, "volumemute"   : 0
, "webradio"     : false
, "display"          : false
, "audiocd"          : false
, "camilladsp"       : false
, "color"            : ""
, "dabradio"         : false
, "equalizer"        : false
, "multiraudio"      : true
, "relays"           : false
, "screenoff"        : false
, "snapclient"       : false
, "volumenone"       : true
}
X, "repeat" : false                             <----  ERROR Here!
, "random" : false
, "single" : false
, "file"      : ""
, "pllength"  : 0
, "song"      : 0
, "state"     : "stop"
, "timestamp" : 1686057287484
, "coverart" : ""
, "hostname" : "raoffice.cmhomenet.lan"
, "ip"       : "192.168.86.31"
, "login" : false }```

EDIT1

Even with:

- a brand new SD card 
- flashed to i20230521
- updated to 20230603
- configured as in https://github.com/rern/rAudio/discussions/1308#discussioncomment-6040688
- applied sed command from rern above
- DEFINITELY did not enable rAudio Server
- 
I am getting the JSON error noted above again - why ?

EDIT 2

It appears that when the Shared Data area is created the file /srv/http/data/system/display.json is being copied (moved ?) to /srv/http/data/backup/display.json and a new version is not created in /srv/http/data/system. Copying the display.json file back to /srv/http/data/system (seems ?) remove the json error noted above.

Chris

**Comment by Sagnasty** [comment]
Fixed in 20230609 - Thanks Rern

Jim
